app/src/main/java/org/coepi/android/cen/CenRepo.kt

```diff
@@ -186,38 +183,54 @@ class CenRepoImpl(
     }
 
     // lastCENKeys gets the last few CENKeys used to generate CENs by this device
-    fun lastCENKeysHex(lim : Int) : String? {
+    private fun lastCENKeysHex(lim : Int) : String? {
         val CENKeys = cenkeyDao.lastCENKeys(lim)
-        if (CENKeys.size > 0) {
+        if (CENKeys.isNotEmpty()) {
             val CENKeysStrings = CENKeys.map{ k -> k.key }
             return CENKeysStrings.joinToString(",")
         }
         return null
     }
 
-    fun periodicCENKeysCheck() {
-        val call = cenkeysCheck(lastCENKeysCheck)
+    private fun periodicCENKeysCheck() {
+        val lastRealmCheck= lastCENKeysCheck.lastCENKeysCheck(1)
+        val curtime = curTimeStamp()
+        var timelastCheck= curtime
+        if( lastRealmCheck.isNotEmpty() ){
+            timelastCheck = lastRealmCheck.last().timestamp
+        }//else{ //never run any check         }*/
+        //insert last time we've checked
+        lastCENKeysCheck.insert(curtime)
+        val call = cenkeysCheck(timelastCheck)
         call.enqueue(object :
             Callback<List<String>> {
             override fun onResponse(call: Call<List<String>?>?, response: Response<List<String>>) {
                 val statusCode: Int = response.code()
-                Log.i("CENApi","${statusCode}");
+                Log.i("CENApi","${statusCode}")
                 if ( statusCode == 200 ) {
                     val r: List<String>? = response.body()
                     r?.let {
-                        var keyMatched=Vector<String>()
+                        val keyMatched=Vector<String>()
+                        val processed =HashSet<String>()
                         for ( i in it.indices ) {
                             it[i]?.let { key ->
-                                val matched = matchCENKey(key, lastCENKeysCheck)
-                                log.i("Trying: ${key}");
-                                if( matched!= null && matched.isNotEmpty() ){
-                                    keyMatched.add(key);
+                                if( !processed.contains(key) ){
+                                    val matched = matchCENKey(key, timelastCheck)
```

Comment:
Why pass `timelastCheck` as max timestamp to `matchCENKey`? Don't we want to match until "right now"?

Reply:
from the server currently arrives a lot of keys, quite always the same, so it seems the server is returning all the keys and the app takes some time to start responding to user actions due to this. Then i saw the lastCenKeyCheck has value 0:
    // last time (unix timestamp) the CENKeys were requested
    var lastCENKeysCheck = 0
So I thought that the server was returning all the keys because they've a timestamp >0, that is the parameter timestamp} to API /cenkeys/timestamp is a "since", and so I thought it was missing this saving. I've asked in chat if this can be saved and nobody has objected. 
In go server specs it's written that server should return all keys in the hour before the parameter, https://github.com/Co-Epi/coepi-backend-go
Returns CENKeys from <timestamp> to <timestamp-3600> (Under consideration)
so if it'll be considered [now-1h, now] passing 0 won't return anything, or maybe the server will check a value > 0

Reply:
I get that it was a (possible) problem in the existing implementation. Just trying to clarify as this part of the code is a bit "unexplored".

You implemented the `from` optimization, to request only the keys since the last update. That's fine.

I'm just wondering why we pass this also as `maxTimestamp` to `matchCENKey`.

Also, looking at the api doc:
```
Returns CENKeys from <timestamp> to <timestamp-3600>
```
I'm not sure how this is meant. I'd expect to be able to pass the last timestamp where I fetched the keys (like you're doing) and it will return everything between that timestamp and now?

Concerning the database, I think someone suggested to you using the preferences. You're also not clearing old values when you insert a new one. We only need to store one timestamp. The preferences are a better fit (you just replace the value each time).